### DISCUSSION TITLE
Automatically Posting to Bluesky on New RSS Items

### DISCUSSION THREAD

**Discussion by giscus** [cfjedimaster/raymondcamden2023#44] [Comments]
# Automatically Posting to Bluesky on New RSS Items

A look at automating posts to Bluesky when new data is posted to a RSS feed.

https://www.raymondcamden.com/2024/11/05/automatically-posting-to-bluesky-on-new-rss-items

**Comment by chrisdrakeuk** [comment]
I'm getting an error with the steps.trigger.event.link being undefined (even though the link and title are visible when hovering over the code).
I expect the fix will be something terribly obvious but I'm no coder...  ;-)

**Reply by chrisdrakeuk** [reply]
The event.link error seems to have solved itself - I substituted an environmental variable for my password, deployed it and ... debug now says: "generateText step is missing or did not return a value".

**Reply by cfjedimaster** [reply]
Care to share a screenshot of your workflow so far?

**Reply by chrisdrakeuk** [reply]
To get around a couple of errors, I've accepted AI debugger suggestions but don't think they've caused the latest problem.

![Pipedream 1](https://github.com/user-attachments/assets/e17def22-bc56-43a2-b4cf-cc61a6ecf6d5)
![Pipedream 2](https://github.com/user-attachments/assets/cd29a6f6-33f9-4bc6-8348-f42e4e59bb4d)

**Reply by cfjedimaster** [reply]
Hmm, hard to figure out from this. If you test the entire workflow, ensuring each step works, do you still get an error saying the link is undefined? As a warning, I'm not going to be around much the rest of the day.

**Reply by chrisdrakeuk** [reply]
The trigger seems to be working okay but the flow stops at generateText:
![Pipedream 3](https://github.com/user-attachments/assets/81561d6f-06d4-4f31-9a0c-99236376b938)

**Reply by chrisdrakeuk** [reply]
This is an example result from the trigger:
![Pipedream 4](https://github.com/user-attachments/assets/188f1368-cad6-48d6-b8a4-7853e24f9aab)

I'd appreciate any hints whenever you're free but this project has been in the background for a month so no panic at my end!

**Reply by cfjedimaster** [reply]
Right now it looks your flow isn't set up right. You've got the trigger, but two shots above, the second text is trying to use the generateText step which isn't in your blog. I'd maybe start fresh, and ensure you get every step in right. Does that make sense?

Looking at my blog post, it definitely does assume some previous knowledge with PD. I've got two options for you - 

a) If you can share your workflow with me, I can try fixing it. That would mean trusting me - also, I don't know if the free tier of PD supports that.
b) I'd be willing to create a video for the blog post that shows it a bit more clearly. However, I can't do this till Friday at the earliest - I'm traveling to a conference.

**Reply by benmillett** [reply]
I ran into the same issue with generateText. I found that I could indeed post if I use a string for Text and comment out the embed (just for troubleshooting, I know that's not the desired outcome). It appears that steps.generateText is the bad step. Is that in a previous part of your workflow, Raymond? (similar to chrisdrakeuk, I can wait; enjoy your conference) 

```
    let rt = new RichText({
//      text: steps.generateText.$return_value.text
        text: "Test from the API"
    });
    await rt.detectFacets(agent);

    await agent.post({
      text:rt.text,
      facets: rt.facets, 
      langs:['en-US'],
      createdAt: new Date().toISOString()
      // ,
      // embed: {
      //   $type: "app.bsky.embed.external",
      //   external:card
      // }
    });
```

**Reply by benmillett** [reply]
I got it working. I think part of the problem for me was Cheerio. Instead, I switched to using the description and the image url directly from the RSS feed item. I also remove any the generateText line and put in my own text.

```
    // let req = await fetch(steps.trigger.event.link);
    // let html = await req.text();
    // let $$ = cheerio.load(html);
    // card.description = $$('meta[name="og:description"]').attr('content');
    // let image = $$('meta[name="og:image"]').attr('content');

    card.description = steps.trigger.event.description.replace(/<[^>]*>/g, '');
    let image = steps.trigger.event.enclosures[0].url;
    
    if (image) {
      let blob = await fetch(image).then(r => r.blob());
      let { data } = await agent.uploadBlob(blob, { encoding:'image/jpeg'} );
      card.thumb = data.blob;
    } else {
      card.thumb = null;
    }
  
    let rt = new RichText({
//      text: steps.generateText.$return_value.text
        text: `New on the blog: "${steps["trigger"]["event"]["title"]}"`
        
    });
    await rt.detectFacets(agent);
```

**Reply by chrisdrakeuk** [reply]
> a) If you can share your workflow with me, I can try fixing it. That would mean trusting me - also, I don't know if the free tier of PD supports that. b) I'd be willing to create a video for the blog post that shows it a bit more clearly. However, I can't do this till Friday at the earliest - I'm traveling to a conference.

I think [this link](https://pipedream.com/@ceedee/invite?token=d78c74be307af70114ee762e37b0361c) invites you to join my workspace.

I've made the changes that benmillett detailed and the test run successfully posted to my BlueSky account.

Whenever you get the free time, feel free to make any changes that may result in the (sporadic but at least daily) source posts appearing on BS.

And thank you for getting me started on a new 'adventure' - it's feeling a lot like my first: pulling apart 6502 code in '74! ;-)

**Reply by chrisdrakeuk** [reply]
Thank you so much for joining my plea - I was beginning to doubt my efforts...

> I got it working. I think part of the problem for me was Cheerio. Instead, I switched to using the description and the image url directly from the RSS feed item. I also remove any the generateText line and put in my own text.
> 
> ```
>     // let req = await fetch(steps.trigger.event.link);
>     // let html = await req.text();
>     // let $$ = cheerio.load(html);
>     // card.description = $$('meta[name="og:description"]').attr('content');
>     // let image = $$('meta[name="og:image"]').attr('content');
> 
>     card.description = steps.trigger.event.description.replace(/<[^>]*>/g, '');
>     let image = steps.trigger.event.enclosures[0].url;

The debugger insisted I needed to define _steps.trigger.event.description_ and _steps.trigger.event.enclosures_ 
```
    card.description = steps.trigger.event.description 
      ? steps.trigger.event.description.replace(/<[^>]*>/g, '') 
      : '';
    
    let image = steps.trigger.event.enclosures && steps.trigger.event.enclosures.length > 0 
      ? steps.trigger.event.enclosures[0].url 
      : null;
```
 but eventually the test posted:

> New post: "${steps["trigger"]["event"]["title"]}

to my BS account.

I count this as a pretty dramatic step forward, not that I have a clue if it means a new rss post will appear when the trigger next fires.

**Reply by chrisdrakeuk** [reply]
Unfortunately the trigger only results in "New post: "${steps["trigger"]["event"]["title"]}" getting posted...

**Comment by gringier** [comment]
Hey Raymond, It isn't working for me either. 

Would you (or someone who got these steps to work) consider making a quick video that non-coders can follow?

Thanks.

**Reply by cfjedimaster** [reply]
I 100% plan on doing a video, but not till Friday morning at the earliest.

**Comment by cfjedimaster** [comment]
Hey everyone, I posted a video and I hope this helps. I literally just posted so it may be encoding still, but let me know if this doesn't help and I'll keep on trying. ;) 

https://www.youtube.com/watch?v=_yp9U-QJgOM

**Comment by ghost** [comment]
Okay... I've done everything the video said, even copy/pasted the GitHub code, but I'm still getting the following:

> The error occurs because steps.trigger.event.link is undefined, leading to a failure when trying to parse the URL. This happens on line 35 of your code. Ensure that the link field in the trigger event is correctly populated.

What's odd is when I hold my mouse over the below I see what I'm expecting to see

>  uri:steps.trigger.event.link,
>  title:steps.trigger.event.title,

...but where it gets odd is it says Line 35, which is actually this:

> let blob = await fetch(image).then(r => r.blob());

This is most confusing as I'm not talented in this arena but your methodology looks like the easiest so I'd love to learn it. I also heard you mention Mastodon which I'm also interested in :)

Help? :)

**Reply by cfjedimaster** [reply]
So when you test, are you executing each step? Ie, you need to ensure the trigger runs so it has a value. Once you execute the trigger, then try running the next step.

**Reply by ghost** [reply]
Yes. So. "trigger" is green with the following:
> Success
> Reference exports in future steps via the **steps** object
And the Exports section gives me with metadata from the first entry from my RSS feed

Then, "generateText" is green with the following:
> Success
> Reference exports in future steps via the **steps** object
And the Exports section gives me text first entry from my RSS feed as requested

Then, "posttoBluesky" is red. I went back and tried Test Workflow which returns all the same hoopla. Is this because I'm on the free version?

**Reply by cfjedimaster** [reply]
Can you share the RSS you are using,  and the URL of the item you are testing with. Do you have an og:image meta tag?

**Reply by ghost** [reply]
That's a good idea...
 RSS Feed
> https://feed.podbean.com/thetoothless/feed.xml

URL being tested
> https://thetoothless.podbean.com/e/01-determined-the-toothless-philosopher/

IMG
> &lt;meta property="og:image" content="https://d2bwo9zemjwxh5.cloudfront.net/ep-logo/pbblog18340650/FreeWill_Determined_pvrkpr_1200x628.jpg?s=2cc1c23f5ea6c8a97b85ebb80324a6ca&amp;e=png"&gt;

**Reply by cfjedimaster** [reply]
Looking at this now.

**Reply by cfjedimaster** [reply]
Ok, I found it. Your meta tag is:

<meta property="og:image" content="https://d2bwo9zemjwxh5.cloudfront.net/ep-logo/pbblog18340650/FreeWill_Determined_pvrkpr_1200x628.jpg?s=2cc1c23f5ea6c8a97b85ebb80324a6ca&amp;e=png">

Notice it is property="og:image", not name. Looking online, it looks like this might more appropriate than what I used. While my code is working for my site (using another tool to test the social media preview), I think I need to change to what you have. 

Now luckily, it's super easy for you to fix.  Find this line:

    let image = $$('meta[name="og:image"]').attr('content');

And add this after: if(!image) image = $$('meta[property="og:image"]').attr('content');

Please confirm, and I'll update the blog post, and just use a note in the Youtube description.

**Reply by ghost** [reply]
Sweet! A new error!
> Invalid app.bsky.feed.post record: Record/embed/external must have the property "description"

If I Debug with AI it says this:
> The error occurs because the card object is missing the description property, which is required for the app.bsky.feed.post record. Ensure that the description is correctly extracted and assigned.

I have two potential meta tags for this:
> &lt;meta property="og:description" content="This is me discussing why I&#39;m interested in free will. I intended it to be primarily Robert Sapolsky&#39;s book &quot;Determined&quot;, but I had to wrap in Sam Harris&#39; book &quot;Free Will&quot;. This will be part of an ongoing review of - probably - both books. Hope you enjoy."&gt;
> &lt;meta name="description" content="This is me discussing why I&#39;m interested in free will. I intended it to be primarily Robert Sapolsky&#39;s book &quot;Determined&quot;, but I had to wrap in Sam Harris&#39; book &quot;Free Will&quot;. This will be part of an ongoing review of - probably - both books. Hope you enjoy."&gt;

Not sure I can use the same method from "let image" on "card.description"...

**Reply by ghost** [reply]
By the way, I really appreciate you taking the time to help me out with this. I'm definitely not trying to be difficult.

**Reply by benmillett** [reply]
try adding this after the card.description line:
        if(!card.description) card.description = $$('meta[property="og:description"]').attr('content');

That worked for me. (I commented out the changes I made in an earlier thread to test this. I do get the description now.)

**Reply by ghost** [reply]
OH THAT DID WORK! I wasn't sure I could pull that off... This is what that looks like now. ALL GREEN!

>    card.description = $$('meta[name="og:description"]').attr('content');
>    if(!card.description) card.description = $$('meta[property="og:description"]').attr('content');    
>    let image = $$('meta[name="og:image"]').attr('content');
>    if(!image) image = $$('meta[property="og:image"]').attr('content');

**Reply by ghost** [reply]
Okay... Mastodon? Do you have those instructions somewhere, too? :)

**Reply by benmillett** [reply]
Pipedream has a Mastodon module in it already, <a href="https://pipedream.com/apps/mastodon/actions/post-status">post_status</a>. I've got that sitting between my RSS feed and Bsky module.

This is what I put into the status box of the form:
`New on the blog: {{steps.trigger.event.title}} {{steps.trigger.event.link}}`

**Reply by cfjedimaster** [reply]
Thank you @benmillett  - looks like both my meta tags 'work', but should be property. I'm going to change that code in my blog post too. I've got a meeting, but updates on the post should be soonish. 

Mastodon - this is older, but still works: https://www.raymondcamden.com/2022/12/06/automatically-posting-to-mastodon-and-twitter-on-new-rss-items 

Also, if you search on here for Mastodon, it's all bot related.

**Reply by cfjedimaster** [reply]
Oh heck yea, use that. They didn't have it when I built my first integrations. Dang Pipedream making things better.

**Reply by cfjedimaster** [reply]
Just as a word of warning, I think property should be used everywhere - so my code (I mean the code I publish to GitHub, and in the blog post after editing), will ONLY look for property. I'll have some 'notes' on top to explain the recent changes.

**Reply by ghost** [reply]
@benmillett & @cfjedimaster

Red on the Mastodon! Aren't you glad you helped me :)
> AxiosError - Request failed with status code 404

I also can't figure out how to remove the Connection... #amitheidiot

**Reply by ghost** [reply]
I figured it out. Thank you both SO very much!

**Comment by cfjedimaster** [comment]
Per the comments by @toothless-666 and @benmillett, I've added a note to the top, and edited the code in here. Thanks everyone!

**Comment by fallingrock** [comment]
Well, I thought your approach was going to work great, but I ran out of work credits on pipedream almost immediately.

**Reply by cfjedimaster** [reply]
Did you try tweaking the RSS trigger timing?

**Reply by fallingrock** [reply]
It was scheduled to run every hour.

I only had 10 credits.

**Reply by cfjedimaster** [reply]
I'd tweak that to every 4, or maybe even 6 hours. I haven't looked at the free tier in PD in a while, but surely it would support that (I hope!). That being said, there _are_ other options out there where you could borrow my code and apply it there as well. If there's enough interest, I could demonstrate how to do this w/ a Cloudflare Scheduled worker.

**Reply by cfjedimaster** [reply]
Yea, so looking at PD pricing, the free tier is 10 credits per day, so once an hour would kill that. Unless you blog like crazy, you can really crank it down.

**Comment by StefanGlasUE** [comment]
Hey Raymond!

I wanted to try your method, but Pipedream does not offer that "New Item in RSS Feed" anymore. At least it does not appear for me. Am I blind? Do you have any other idea for posting an RSS-Feed automatically to Bluesky?

Thanks in adcance for your help!

**Reply by cfjedimaster** [reply]
It's possible it's not in the free tier? I also know I've had issues with their search before. Worst comes to worst, I'd ask on their Slack.
